### PR TITLE
chore(docs): Created `GitHub runner cryptographic overview`

### PR DESCRIPTION
**Pull request by alithethird**
Applicable spec: SEC0030 - SSDLC - Cryptographic Documentation

### Overview
Created the `GitHub runner cryptographic overview` document.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied.
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied.
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`.
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`).
- [x] The changelog is updated with changes that affects the users of the charm.

### CODE REVIEW

File: docs/reference/cryptographic-overview.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# GitHub runner cryptographic overview
+This document provides an overview of the cryptographic technologies used in the GitHub runner charm, including encryption/decryption, hashing, and digital signatures.
+
+## Overall Description of Cryptographic Technology Use
```
Comment (arturo-seijas) [inline docs/reference/cryptographic-overview.md:4]: In general, simpler grammar is preferred for technical documentation to lower the barrier for comprehension, for instance, some suggestions
s/that need to be executed/to be executed
s/retrieve information about how the repositories are set up/retrieve the repositories' configuration
s/the runner creates virtual machines in OpenStack and uses these machines to run the workload/the runner creates virtual machines in OpenStack to run the workload

Would you mind double checking if you can further simplify?

Comment (alithethird) [reply docs/reference/cryptographic-overview.md:4]: Sure can do! Thanks for the tip!

File: docs/reference/cryptographic-overview.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+### Docker Hub Cache
+[Docker Hub cache](https://github.com/canonical/docker-registry-charm) connection is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
+### Aproxy
+[Aproxy](https://github.com/canonical/aproxy) supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.
+## Packages or Technology Providing Cryptographic Functionality
+The following packages or technologies provide cryptographic functionality:
+- [OpenSSL](https://www.openssl.org/) library is being used.
+- For TLS and HTTPS connections urllib3 python library is used.
+- Default clients in Ubuntu for SSH and TLS are being used.
+- [Python hashlib](https://docs.python.org/3/library/hashlib.html) package is being used for sha256 checksum calculation/verification of runner binary and Cloud Init.
```
Comment (erinecon) [inline docs/reference/cryptographic-overview.md:45]: Nitpick: Does sha256 need to be capitalised?

Comment (alithethird) [reply docs/reference/cryptographic-overview.md:45]: When I search for it online it is always capitalized but still wrong I guess. I see it as `SHA-256` in a lot of places.

Ex.: [Wikipedia](https://en.wikipedia.org/wiki/SHA-2), [SimpliLearn](https://www.simplilearn.com/tutorials/cyber-security-tutorial/sha-256-algorithm)

File: docs/reference/cryptographic-overview.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# GitHub runner cryptographic overview
+This document provides an overview of the cryptographic technologies used in the GitHub runner charm, including encryption/decryption, hashing, and digital signatures.
+
+## Overall Description of Cryptographic Technology Use
+The GitHub runner charm uses various cryptographic technologies to ensure secure communication and data integrity. The following sections describe each of the components.
```
Comment (yanksyoon) [inline docs/reference/cryptographic-overview.md:5]: Should we have a new line to separate the headers? :D

File: docs/reference/cryptographic-overview.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+### TLS
+The GitHub runner communicates with GitHub to receive information about the workflows to be executed, retrieve the repositories config and send back workflow logs to the user. This communication uses [urllib3](https://urllib3.readthedocs.io/en/stable/) under the hood using TLS 1.3.
+
+In OpenStack mode, the runner creates virtual machines to run the workload in OpenStack. The GitHub runner charm interacts with the OpenStack API to create and manage virtual machines. These interactions are secured via TLS. The communication between these virtual machines and runner is done via SSH and secured by 256 byte keypairs.
```
Comment (yanksyoon) [inline docs/reference/cryptographic-overview.md:9]: Nit: would the following info be helpful?

The SSH key is created by OpenStack, which uses ssh PEM keys (OpenSSL).
```
$ ssh-keygen -l -f ./<keyfile>
2048 SHA256:w0gf4CC2xY3X9z+WRwT1xh4osy1j25hsm3fCg7bUOqk no comment (RSA)
```

**Review by yanksyoon — DISMISSED** [review]
wow, great research, thank you for doing it!

File: docs/reference/cryptographic-overview.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+[DockerHub charm](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:11]:
```suggestion
[Docker Registry](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
```

File: docs/reference/cryptographic-overview.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+Images that run in the OpenStack vm are built using the [Image Builder Charm](https://github.com/canonical/github-runner-image-builder-operator). This charm needs to download the runner binary, cloud image to base the image on and yq. All these images are downloaded with TLS.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:13]: nit: Technically its not the charm, but the application managed by the charm, so maybe rephrase to "This charm manages an application that needs to download the runner binary..." (cc @yanksyoon )

File: docs/reference/cryptographic-overview.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+The GitHub runner charm supports being deployed behind an HTTP proxy. [Aproxy](https://github.com/canonical/aproxy) is installed and enabled when an HTTP proxy is detected so that jobs executing in the runner VMs don’t have to configure the proxy themselves. Aproxy is a transparent proxy service for HTTP and HTTPS/TLS connections. Aproxy works by pre-reading the Host header in HTTP requests and SNI in TLS hellos; it forwards HTTP proxy requests with the hostname therefore, complies with HTTP proxies requiring destination hostname for auditing or access control. Aproxy doesn't and can't decrypt the TLS connections. It works by reading the plaintext SNI information in the client hello during the TLS handshake, so the authentication and encryption of TLS are never compromised. Aproxy supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.
+### Signature Verification
+Cloud image is verified by SHA256 checksum. Runner binary is also downloaded by [GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:17]: Could we elaborate which Cloud image is meant here?

File: docs/reference/cryptographic-overview.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+### User SSH Access
+Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:19]:
```suggestion
Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses the SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.
```

File: docs/reference/cryptographic-overview.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+### TLS
+- Communication with GitHub is done via TLS v1.3
+- [The Repo Policy Compliance tool](https://github.com/canonical/repo-policy-compliance) communicates with the [GitHub API](https://docs.github.com/en/rest?apiVersion=2022-11-28) using TLS v1.3.
+- Communication with the OpenStack API uses TLS v1.3.
+### Signature Verification
+Cloud images are verified by SHA256 checksum. Runner binary is also downloaded by [the GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
+### User SSH Access
+Tmate secures SSH connections using [OpenSSL](https://www.openssl.org/).
+### RSA
+The communication between these virtual machines and runner is done via SSH and secured by RSA 256 byte keypairs.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:31]:
```suggestion
The communication between these virtual machines and rthe unner is done via SSH and secured by RSA 256 byte keypairs.
```

File: docs/reference/cryptographic-overview.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+## Cryptographic Technology Exposed to Users
+The following sections describe the cryptographic technologies exposed to the user:
+### Tmate
+- [Tmate](https://tmate.io/) uses RSA (384 byte) and ED25519 (32 byte) fingerprints for connections from users to [Tmate](https://tmate.io/) and from the managers to [Tmate](https://tmate.io/). [OpenSSL](https://www.openssl.org/) is being used by [Tmate](https://tmate.io/) to secure the connection between user/manager to runner.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:35]:
```suggestion
- [Tmate](https://tmate.io/) uses RSA (384 byte) and ED25519 (32 byte) fingerprints for connections from users to [Tmate](https://tmate.io/) and from the managers to [Tmate](https://tmate.io/). [OpenSSL](https://www.openssl.org/) is being used by [Tmate](https://tmate.io/) to secure the connection between the user/manager to the runner.
```

**Review by cbartz — COMMENTED** [review]
Thanks! Mostly have nits

File: docs/reference/cryptographic-overview.md [hunk 11]
```diff
@@ -3,45 +3,58 @@ This document provides an overview of the cryptographic technologies used in the
...
 
-[DockerHub charm](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
+[DockerHub Registry](https://charmhub.io/docker-registry) is used as a cache between the official DockerHub and GitHub runners to avoid rate limiting. Communication between GitHub runners and DockerHub cache is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
 
-Images that run in the OpenStack vm are built using the [Image Builder Charm](https://github.com/canonical/github-runner-image-builder-operator). This charm needs to download the runner binary, cloud image to base the image on and yq. All these images are downloaded with TLS.
+Images that run in the OpenStack vm are built using the [Image Builder](https://github.com/canonical/github-runner-image-builder-operator). This application needs to download the runner binary, yq and the cloud image to base the image on. All these images are downloaded with TLS.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:14]:
```suggestion
Images that run in the OpenStack vm are built using the [Image Builder](https://github.com/canonical/github-runner-image-builder. This application needs to download the runner binary, yq and the cloud image to base the image on. All these images are downloaded with TLS.
```

File: docs/reference/cryptographic-overview.md [hunk 12]
```diff
@@ -3,45 +3,58 @@ This document provides an overview of the cryptographic technologies used in the
...
 The following cryptographic technologies are used internally by our product:
+
 ### TLS
 - Communication with GitHub is done via TLS v1.3
 - [The Repo Policy Compliance tool](https://github.com/canonical/repo-policy-compliance) communicates with the [GitHub API](https://docs.github.com/en/rest?apiVersion=2022-11-28) using TLS v1.3.
 - Communication with the OpenStack API uses TLS v1.3.
+
 ### Signature Verification
-Cloud images are verified by SHA256 checksum. Runner binary is also downloaded by [the GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
+Cloud images that are downloaded by the Image Builder are verified by SHA256 checksum. Runner binary is also downloaded by [the GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
```
Comment (cbartz) [inline docs/reference/cryptographic-overview.md:33]: @yanksyoon Can you have a look on "Cloud images that are downloaded by the Image Builder are verified by SHA256 checksum." if its correct in terms of the technical implementation?

File: docs/reference/cryptographic-overview.md (hunk 7 above)
Comment (cbartz) [reply docs/reference/cryptographic-overview.md:19]: I think this sentence should be updated here, too.

**Review by cbartz — COMMENTED** [review]
Thanks. Some follow-up comments.

**Comment by cbartz** [comment]
@alithethird Whats the status here, please?

File: docs/reference/cryptographic-overview.md (hunk 12 above)
Comment (alithethird) [reply docs/reference/cryptographic-overview.md:33]: @yanksyoon Is this technically correct?

File: docs/reference/cryptographic-overview.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+### Signature Verification
+Images that run in the OpenStack vm (Cloud images) are verified by SHA256 checksum. Runner binary is also downloaded by [GitHub Runner Charm](https://github.com/canonical/github-runner-operator) and verified by SHA256 in this charm.
+
+### User SSH Access
+Sometimes users need to access the VM instance that is running the workload, to establish this connection [Tmate](https://tmate.io/) is used. Tmate uses the SSH protocol to secure shell connections between users and the GitHub runner. The connection is secured with rsa keypair and ed25519 fingerprints.
```
Comment (erinecon) [inline docs/reference/cryptographic-overview.md:22]: Nitpick: Capitalise RSA (I see it's capitalised elsewhere in the document, so for consistency it should be capitalised here)

File: docs/reference/cryptographic-overview.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+[Docker Hub cache](https://github.com/canonical/docker-registry-charm) connection is secured via TLS 1.3 and certified by [Let’s Encrypt](https://letsencrypt.org/).
+
+### Aproxy
+[Aproxy](https://github.com/canonical/aproxy) supports TLS 1.0 and above except TLS 1.3 Encrypted Client Hello.
+
+## Packages or Technology Providing Cryptographic Functionality
+The following packages or technologies provide cryptographic functionality:
+- [OpenSSL](https://www.openssl.org/) library is being used.
+- For TLS and HTTPS connections urllib3 python library is used.
+- Default clients in Ubuntu for SSH and TLS are being used.
```
Comment (erinecon) [inline docs/reference/cryptographic-overview.md:57]: Hi, is it possible to give a brief explanation for what these libraries are being used for?

File: docs/reference/cryptographic-overview.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+- [Python hashlib](https://docs.python.org/3/library/hashlib.html) package is being used for SHA256 checksum calculation/verification of runner binary and Cloud Init.
+- OpenStack client is being used to generate keypairs.
+- [Aproxy](https://github.com/canonical/aproxy) using the [golang.org/x/crypto](http://golang.org/x/crypto) package.
```
Comment (erinecon) [inline docs/reference/cryptographic-overview.md:60]: Is it possible to give a brief explanation for what Aproxy is being used for?

**Review by weiiwang01 — APPROVED** [review]

File: docs/reference/cryptographic-overview.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
+# GitHub runner cryptographic overview
+This document provides an overview of the cryptographic technologies used in the GitHub runner charm, including encryption/decryption, hashing, and digital signatures.
```
Comment (erinecon) [inline docs/reference/cryptographic-overview.md:2]: Two nitpicks on this sentence: 
1. I think you can separate encryption and decryption with a comma 
2. The style guide recommends not using the Oxford comma, so remove the comma after "hashing"

**Review by erinecon — APPROVED** [review]
Looks great to me, thanks! Once @yanksyoon responds to the unresolved comment I say it's good to go

File: docs/reference/cryptographic-overview.md (hunk 12 above)
Comment (yanksyoon) [reply docs/reference/cryptographic-overview.md:33]: Yup :)
https://github.com/canonical/github-runner-image-builder/blob/main/src/github_runner_image_builder/cloud_image.py#L153

Comment (alithethird) [reply docs/reference/cryptographic-overview.md:33]: Thanks for confirming!
